### DISCUSSION TITLE
Usage of modern packaging tools

### DISCUSSION THREAD

**Discussion by oscarmcm** [django-wiki/django-wiki#1226] [Ideas]
This is more a continuation of the work to be made in #1199 so I have a few questions:

Should we update our GitHub Actions to use [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) in order to produce more Python wheels for different platforms.

The other thing is that we need to modernize the build/release of the project, and now there's a few package managers out there. And after reviewing some of them, I feel that [hatch](https://github.com/pypa/hatch) is the go-to solution as this is focused only in the packaging side, plus is backed by the PyPA group and plays nicely with other tools such as `twine` and `build`.

Does that sounds fine? Any other comments or suggestions will be appreciated.

**Comment by benjaoming** [comment]
Sounds good re: hatch! Am excited to see how it works :tada: 

> Should we update our GitHub Actions

We actually don't have any GA :) I've been happily investing my time in alternatives to GitHub Actions, and when Travis died, we just went straight to Circle CI. It's pretty cool for advanced Docker stuff, but that's not something that we are doing here.. django-wiki is just basic stuff.

Regarding releases to PyPi, I've been shipping them from a local environment and signing them with my own keys (even though that's largely unsupported by PyPi). Before a release, I do a release PR. I'm happy to continue uploading releases manually, but if we can get to a step where the entire release is built and uploaded via Circle CI, that's great, too! I like the advantage of having many people able to do releases + doing reproducible builds.

There's some stuff about preparing releases, i.e. Transifex and asset building, that you can refer to in the Makefile. I dunno if we want a "prepare release" step separate somehow from the "upload release" step in order to be able to verify releases. What could worry me would be that we release some garbage from Transifex or that asset building didn't work.

Btw Circle CI has [this example](https://circleci.com/developer/orbs/orb/pennlabs/pypi-tools), it seems basic, but perhaps possible to add the hatch commands necessary? What do you make of it?

**Reply by oscarmcm** [reply]
Cool! I'll use hatch to port and remove the old `setup.py` and `setup.cfg` to the new `pyproject.toml` and in the same manner try to port some commands from the Makefile into the [scripts section for hatch](https://hatch.pypa.io/latest/config/environment/overview/#scripts)

The first proposal to it will be get off from `setup.py|.cfg` then, maybe in another issue we can focus on building multiple wheels and automate the release process.

**Reply by benjaoming** [reply]
Bravo, sounds like the perfect first step -- and perhaps also the more important part! Thanks, looking forwards to seeing it :)

**Comment by oscarmcm** [comment]
We just landed support for `pyproject.toml` via PR #1227 

I'll keep this open so we can discuss if we can migrate to [ruff](https://github.com/charliermarsh/ruff) and drop the usage of `black` and `flake8`

Feel free to keep the discussion here with more tools and things that we should modernize.

**Reply by benjaoming** [reply]
Sounds great re:ruff - but I think that it isn't intended to replace Black. This is in their README:

> As a project, Ruff is designed to be used alongside Black and, as such, will defer implementing stylistic lint rules that are obviated by autoformatting.

**Reply by benjaoming** [reply]
The progress made in #1227 is really outstanding, great work :raised_hands:
